File: src/ratfor/hoeffd.r
#  ratfor -o ../hoeffd.f hoeffd.r
#
SUBROUTINE hoeffd(xx, n, p, dmat, aadmat, madmat, npair, x, y, rx, ry, rj)
IMPLICIT DOUBLE PRECISION (a-h,o-z)
INTEGER p, npair(p,p)
DOUBLE PRECISION xx(n,p), dmat(p,p), aadmat(p,p), madmat(p,p), x(n), y(n), rx(n), ry(n), rj(n), maxad

    DO i=1, p {
        np=0
        DO k=1, n {
            if(xx(k,i) < 1d49) np = np + 1
        }
        npair(i,i) = np

        DO j=(i+1),p {
            m = 0
            DO k=1,n {
                xk = xx(k,i)
                yk = xx(k,j)
                if(xk < 1d49 & yk < 1d49) {
                    m = m + 1
                    x(m) = xk
                    y(m) = yk
                }
            }
            npair(i,j) = m
            if(m > 4) {
                CALL hoeff(x, y, m, d, aad, maxad, rx, ry, rj)
                dmat(i,j) = d
                aadmat(i,j) = aad
                madmat(i,j) = maxad
            }
            else dmat(i,j) = 1d50
        }
    }
    DO i=1,p {
        dmat(i,i) = 1d0/30d0
        DO j=(i+1),p {
            dmat(j,i) = dmat(i,j)
            npair(j,i) = npair(i,j)
            aadmat(j,i) = aadmat(i,j)
            madmat(j,i) = madmat(i,j)
        }
    }
    RETURN
END  

SUBROUTINE hoeff(x, y, n, d, aad, maxad, rx, ry, rj)
IMPLICIT DOUBLE PRECISION (a-h,o-z)
DOUBLE PRECISION x(n), y(n), rx(n), ry(n), rj(n), maxad
#	INTEGER iwork(1)
#	CALL rank(n, x, work, iwork, rx)
#	CALL rank(n, y, work, iwork, ry)
    CALL jrank(x, y, n, rx, ry, rj)
    q = 0d0
    r = 0d0
    s = 0d0
    aad = 0d0
    maxad = 0d0
    z = n
    DO i=1,n {
        rxi = rx(i)
        ryi = ry(i)
        rji = rj(i)
        ad  = dabs((rji/z) - (rxi/z)*(ryi/z))
        aad = aad + ad
        maxad = dmax1(maxad, ad)
	q = q + (rxi-1d0)*(rxi-2d0)*(ryi-1d0)*(ryi-2d0)
	r = r + (rxi-2d0)*(ryi-2d0)*(rji-1d0)
	s = s + (rji-1d0)*(rji-2d0)
    }
    aad = aad / z
    d = (q-2d0*(z-2d0)*r+(z-2d0)*(z-3d0)*s)/z/(z-1d0)/(z-2d0)/(z-3d0)/(z-4d0)
    RETURN
END

#	Use C version of this which is much faster (since uses a sort)
#	SUBROUTINE rank(x, n, r)   # simple rank with midranks for ties
#	REAL*4 x(1), r(1)
#	DO i=1,n {
#	  xi=x(i)
#	  ir=2   # will be 2*rank(x(i))
#	  DO j=1,n {
#	    if(i.ne.j) {
#	      if(x(j)<xi) ir=ir+2
#	      if(x(j)==xi) ir=ir+1
#	    }
#	  }
#	  r(i)=float(ir)/2.0
#	}
#	RETURN
#	END

SUBROUTINE jrank(x, y, n, rx, ry, r) # joint rank with midranks for ties
INTEGER n
DOUBLE PRECISION x(n), y(n), rx(n), ry(n), r(n), cx, cy, ri, rix, riy, xi, yi
    DO i=1,n {
        xi  = x(i)
	yi  = y(i)
        ri  = 1d0
        rix = 1d0
        riy = 1d0
        DO j=1,n {
            if(i .ne. j) {
	        cx = 0d0
                if(x(j) < xi) cx = 1d0
                if(x(j) == xi) cx = .5d0
                cy = 0d0
                if(y(j) < yi) cy = 1d0
                if(y(j) == yi) cy = .5d0
                rix = rix + cx
                riy = riy + cy
	        ri  = ri + cx*cy
            }
        }
        rx(i) = rix
        ry(i) = riy
        r(i) = ri
    }
    RETURN
END

File: src/hoeffd.f90
! Converted from Ratfor to Fortran 2018 by Claude AI 2024-10-26
!
! Hoeffding's D Statistics Implementation
! Modern Fortran without module structure

subroutine hoeffd(xx, n, p, dmat, aadmat, madmat, npair, x, y, rx, ry, rj)
  use, intrinsic :: iso_fortran_env, only: real64
  implicit none
  
  ! Arguments
  integer, intent(in) :: n, p
  real(real64), intent(in) :: xx(n,p)
  real(real64), intent(out) :: dmat(p,p), aadmat(p,p), madmat(p,p)
  integer, intent(out) :: npair(p,p)
  real(real64), intent(out) :: x(n), y(n), rx(n), ry(n), rj(n)
  
  ! Local variables
  integer :: i, j, k, m, np
  real(real64) :: xk, yk, d, aad, maxad
  real(real64), parameter :: LARGE_VAL = 1.0e49_real64
  
  ! Initialize arrays
  dmat = 0.0_real64
  aadmat = 0.0_real64
  madmat = 0.0_real64
  npair = 0
  
  do i = 1, p
    np = count(xx(:,i) < LARGE_VAL)
    npair(i,i) = np
    
    do j = i+1, p
      m = 0
      do k = 1, n
        xk = xx(k,i)
        yk = xx(k,j)
        if (xk < LARGE_VAL .and. yk < LARGE_VAL) then
          m = m + 1
          x(m) = xk
          y(m) = yk
        end if
      end do
      
      npair(i,j) = m
      if (m > 4) then
        call hoeff(x, y, m, d, aad, maxad, rx, ry, rj)
        dmat(i,j) = d
        aadmat(i,j) = aad
        madmat(i,j) = maxad
      else
        dmat(i,j) = LARGE_VAL
      end if
    end do
  end do
  
  ! Fill lower triangle
  do i = 1, p
    dmat(i,i) = 1.0_real64/30.0_real64
    do j = i+1, p
      dmat(j,i) = dmat(i,j)
      npair(j,i) = npair(i,j)
      aadmat(j,i) = aadmat(i,j)
      madmat(j,i) = madmat(i,j)
    end do
  end do
  
end subroutine hoeffd

subroutine hoeff(x, y, n, d, aad, maxad, rx, ry, rj)
  use, intrinsic :: iso_fortran_env, only: real64
  implicit none
  
  ! Arguments
  integer, intent(in) :: n
  real(real64), intent(in) :: x(n), y(n)
  real(real64), intent(out) :: d, aad, maxad
  real(real64), intent(out) :: rx(n), ry(n), rj(n)
  
  ! Local variables
  integer :: i
  real(real64) :: q, r, s, z, rxi, ryi, rji, ad
  
  call jrank(x, y, n, rx, ry, rj)
  
  q = 0.0_real64
  r = 0.0_real64
  s = 0.0_real64
  aad = 0.0_real64
  maxad = 0.0_real64
  z = real(n, real64)
  
  do i = 1, n
    rxi = rx(i)
    ryi = ry(i)
    rji = rj(i)
    ad = abs((rji/z) - (rxi/z)*(ryi/z))
    aad = aad + ad
    maxad = max(maxad, ad)
    q = q + (rxi-1.0_real64)*(rxi-2.0_real64)*(ryi-1.0_real64)*(ryi-2.0_real64)
    r = r + (rxi-2.0_real64)*(ryi-2.0_real64)*(rji-1.0_real64)
    s = s + (rji-1.0_real64)*(rji-2.0_real64)
  end do
  
  aad = aad / z
  d = (q-2.0_real64*(z-2.0_real64)*r+(z-2.0_real64)*(z-3.0_real64)*s) / &
      (z*(z-1.0_real64)*(z-2.0_real64)*(z-3.0_real64)*(z-4.0_real64))
  
end subroutine hoeff

subroutine jrank(x, y, n, rx, ry, r)
  use, intrinsic :: iso_fortran_env, only: real64
  implicit none
  
  ! Arguments
  integer, intent(in) :: n
  real(real64), intent(in) :: x(n), y(n)
  real(real64), intent(out) :: rx(n), ry(n), r(n)
  
  ! Local variables
  integer :: i, j
  real(real64) :: xi, yi, ri, rix, riy, cx, cy
  
  do i = 1, n
    xi = x(i)
    yi = y(i)
    ri = 1.0_real64
    rix = 1.0_real64
    riy = 1.0_real64
    
    do j = 1, n
      if (i /= j) then
        cx = merge(0.5_real64, merge(1.0_real64, 0.0_real64, x(j) < xi), x(j) == xi)
        cy = merge(0.5_real64, merge(1.0_real64, 0.0_real64, y(j) < yi), y(j) == yi)
        
        rix = rix + cx
        riy = riy + cy
        ri = ri + cx*cy
      end if
    end do
    
    rx(i) = rix
    ry(i) = riy
    r(i) = ri
  end do
  
end subroutine jrank
